frontend/frontend/protocols/ssh.py

```diff
@@ -57,30 +59,111 @@ def check_channel_shell_request(self, channel: paramiko.Channel) -> bool:
-        # print("modes:", modes.decode("utf-8"))
-        # Allow everything
+        self._session.log_pty_request(term.decode("utf-8"), width, height, pixelwidth, pixelheight)
         return True
 
 
-# todo should be singleton if the class does what it says in the docstring
-class ConnectionManager():
-    """ConnectionManager contains logic for connecting incoming
-    SSH connections to instances of Server"""
-
+class ConnectionHandler(threading.Thread):
     # todo move these to some constants file or something
     CR = b"\r"  # Carriage return (CR)
     LF = b"\n"  # Line feed (LF)
 
+    def __init__(self, transport: paramiko.Transport,
+                 session: logger.SSHSession,
+                 host_key: paramiko.PKey,
+                 usernames: Optional[List[str]],
+                 passwords: Optional[List[str]],
+                 auth_timeout: float) -> None:
+        super().__init__(target=self.handle, daemon=False)
+        self._terminate = False
+        self._usernames = usernames
+        self._passwords = passwords
+        self._host_key = host_key
+        self._transport = transport
+        self._session = session
+        self._auth_timeout = auth_timeout
+        self._lock = threading.Lock()
+
+    def _setup_server(self, host_key: paramiko.PKey,
+                      usernames: Optional[List[str]],
+                      passwords: Optional[List[str]]) -> None:
+        if not self._transport.load_server_moduli():
+            print("Could not load moduli")
+
+        self._transport.add_server_key(host_key)
+        server = Server(self._session, usernames, passwords)
+        self._transport.start_server(server=server)
+
+    def stop(self) -> None:
+        """Stops the `handle` method handling SSH connections"""
+        self._lock.acquire()
+        self._terminate = True
+        self._lock.release()
+
+    # todo check if session.end can be called through decerator
+    def handle(self) -> None:
+        """Waits for an SSH channel to be established and handles
+        the connection"""
+
+        # Setup the server
+        try:
+            self._setup_server(self._host_key, self._usernames, self._passwords)
+        except:
+            # todo log
+            self._session.end()
+            return
+
+        chan = self._transport.accept(self._auth_timeout)
+        if chan is None:
+            # todo log this
+            print("Authentication timeout")
+            self._transport.close()
+            self._session.end()
+            return
+
+        chan.send(
+            b"\r\nWelcome to the Chalmers blueprint server. Please do not steal anything.\r\n")
+
+        # todo wait for user to send shell request
+        # todo don't hardcode
+        chan.settimeout(2)
+        while True:
+            self._lock.acquire()
+            if not self._transport.active or self._terminate:
+                self._lock.release()
+                break
+            self._lock.release()
```

Comment:
Might want to use a try-finally here to guaranteed release the lock. Looks a little dodgy to release inside if.